### PR TITLE
Add reverse routing functionality

### PR DESCRIPTION
**Pull request by alexdavid**
@kevgo initial spike for reverse routing.

Let me know what you think

### CODE REVIEW

File: src/exprestive.coffee [hunk 1]
```diff
@@ -3,6 +3,7 @@ camelCase = require 'camel-case'
 express = require 'express'
 fs = require 'fs'
 path = require 'path'
+{pluralize, singularize } = require 'inflection'
```
Comment (kevgo) [inline src/exprestive.coffee]: Man, this syntax is so cool!

File: src/exprestive.coffee [hunk 2]
```diff
@@ -80,20 +86,35 @@ class Exprestive
     @routesMethod @getRoutesHelperMethods()
 
 
+  # Save a function to @reversePaths to get a url back from a route name
+  registerReverseRoute: ({routeName, url}) ->
+    return unless routeName?
```
Comment (kevgo) [inline src/exprestive.coffee]: Is silently ignoring a missing parameter like this the right thing? If this is an invalid situation, it shouldn't it throw an exception instead.

File: src/exprestive.coffee [hunk 3]
```diff
@@ -80,20 +86,35 @@ class Exprestive
...
-      [ 'GET',    "/#{controllerName}",          'index'   ]
-      [ 'GET',    "/#{controllerName}/new",      'new'     ]
-      [ 'POST',   "/#{controllerName}",          'create'  ]
-      [ 'GET',    "/#{controllerName}/:id",      'show'    ]
-      [ 'GET',    "/#{controllerName}/:id/edit", 'edit'    ]
-      [ 'PUT',    "/#{controllerName}/:id",      'update'  ]
-      [ 'DELETE', "/#{controllerName}/:id",      'destroy' ]
+    singular = singularize controllerName
+    plural = pluralize controllerName
+    for [ httpMethod ,  url                   , controllerAction , routeName          ] in [
```
Comment (kevgo) [inline src/exprestive.coffee]: If you don't pad everything with spaces here, it takes up less horizontal space, and looks more like normal CoffeeScript.

File: src/exprestive.coffee (hunk 1 above)
Comment (alexdavid) [inline src/exprestive.coffee]: Agreed, but I just realized a stray space got in there before the `}`. Leaving a note here for myself to remove it

File: src/exprestive.coffee [hunk 4]
```diff
@@ -80,20 +86,35 @@ class Exprestive
...
+  # Save a function to @reversePaths to get a url back from a route name
+  registerReverseRoute: ({routeName, url}) ->
+    return unless routeName?
+    @reversePaths[camelCase routeName] = (id) ->
+      url.replace ':id', id
+
+
   # A helper method for automatically binding restful controllers in a routes file
   # This is passed as "resources" to the routes file function parameter hash
   resourcesHelperMethod: (controllerName) =>
```
Comment (kevgo) [inline src/exprestive.coffee]: A better name seems `resourcesDirective` or whatever the name is for the elements of a routing file.

File: test/feature/reverse_routing.feature [hunk 5]
```diff
@@ -0,0 +1,31 @@
+Feature: reverse routing
+
+  Background:
+    Given an empty "controllers" directory
+
+
+  Scenario: routes can define reverse routes with 'as'
+    Given a file "routes.coffee" with the contents
+      """
+      module.exports = ({ GET }) ->
```
Comment (kevgo) [inline test/feature/reverse_routing.feature]: This is a good example for where providing the whole routes file makes less sense. What we really care about here is just the `GET` directive. I think the spec should just contain that.

``` cucumber
Scenario: Providing a custom name for routing helpers
  Given the routing definition
    """
    GET '/some/route', to: 'test#index', as: 'foobar'
    """
  Then I have a routing helper "res.locals.paths.foobar" that returns "/some/route"
```

File: test/feature/reverse_routing.feature [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+
+  Scenario Outline: restful routes define reverse routes automatically
+    Given a file "routes.coffee" with the contents
+      """
+      module.exports = ({ resources, GET }) ->
+        resources 'users'
+      """
+    And an exprestive app using defaults
+    Then the function res.locals.<LOCAL> should return "<VALUE>"
```
Comment (kevgo) [inline test/feature/reverse_routing.feature]:
``` cucumber
Then I have a routing helper "res.locals.<LOCAL>" that returns "<VALUE>" when called with argument "<ARGUMENT>"
```

File: test/feature/reverse_routing.feature (hunk 5 above)
Comment (kevgo) [inline test/feature/reverse_routing.feature]: Looking at this spec now, if I don't provide `as:`, then it should create a routing helper named `res.locals.paths.some_route`.

File: test/feature/reverse_routing.feature [hunk 7]
```diff
@@ -0,0 +1,31 @@
+Feature: reverse routing
+
+  Background:
+    Given an empty "controllers" directory
```
Comment (kevgo) [inline test/feature/reverse_routing.feature]: Isn't the controllers directory empty by default?

File: test/feature/support/helpers.coffee [hunk 8]
```diff
@@ -19,12 +19,21 @@ class Helpers
       app.use exprestive(#{optionsStr})
       app.listen #{@port}
 
+      # Route to evaluate functions from tests
+      app.get '/eval/:strToEval', (req, res) ->
```
Comment (kevgo) [inline test/feature/support/helpers.coffee:23]: It would be nice if exprestive allowed to specify flexible routes like that as well.

File: test/feature/support/helpers.coffee [hunk 9]
```diff
@@ -19,12 +19,21 @@ class Helpers
...
+
       # Send a message to parent to let it know the server started successfully
       process.send('server started')
     """
     @createFile 'server.coffee', fileContents, done
 
 
+  createDirectory: (directoryName, done) ->
+    directoryPath = path.join @appPath, directoryName
+    mkdirp directoryPath, done
```
Comment (kevgo) [inline test/feature/support/helpers.coffee]: Merge these two lines

File: test/feature/reverse_routing.feature (hunk 7 above)
Comment (alexdavid) [inline test/feature/reverse_routing.feature]: Yeah, but exprestive throws an error if the controllers directory doesn't exist. That's intended functionality.

Since each test starts with an empty tmp dir there is no controllers directory so this creates one.

File: test/feature/support/helpers.coffee (hunk 8 above)
Comment (alexdavid) [inline test/feature/support/helpers.coffee:23]: Do you mean putting this in an exprestive controller? This would mean creating a routes file for every test run which isn't practical because some test runs define their own custom routes file

Comment (kevgo) [inline test/feature/support/helpers.coffee:23]: I mean that the routes file can contain something like

```
GET '/help/:topic', to: 'help#index'
```

And then if I go to /help/foo it goes to HelpController.index and the `params` hash contains `topic: 'foo'`.

Comment (alexdavid) [inline test/feature/support/helpers.coffee:23]: Yeah it already does that since it uses `express.Router()`

File: test/feature/reverse_routing.feature (hunk 5 above)
Comment (alexdavid) [inline test/feature/reverse_routing.feature]: where would it get `some_route` from? The reverse route shouldn't change when changing the url

File: src/exprestive.coffee (hunk 1 above)
Comment (alexdavid) [inline src/exprestive.coffee]: Done :checkered_flag:

File: src/exprestive.coffee (hunk 2 above)
Comment (alexdavid) [inline src/exprestive.coffee]: Done :checkered_flag:

File: src/exprestive.coffee (hunk 3 above)
Comment (alexdavid) [inline src/exprestive.coffee]: Done :checkered_flag:

File: src/exprestive.coffee (hunk 4 above)
Comment (alexdavid) [inline src/exprestive.coffee]: Done :checkered_flag:

File: test/feature/reverse_routing.feature (hunk 6 above)
Comment (alexdavid) [inline test/feature/reverse_routing.feature]: What do you think about

``` cucumber
Then I have a routing helper "res.locals.<LOCAL>" that returns "<VALUE>"
```

I think this table makes it more clear how it works:

``` cucumber
Examples:
  | LOCAL            | VALUE      |
  | paths.users(123) | /users/123 |
```

than

``` cucumber
Examples:
  | LOCAL       | ARGUMENT | VALUE      |
  | paths.users | 123      | /users/123 |
```

**Comment by alexdavid** [comment]
@kevgo please take another look when you have a chance. I made some changes and added support for passing in a `paths` option.

I also updated the readme with documentation about reverse routing.

File: test/feature/reverse_routing.feature (hunk 7 above)
Comment (kevgo) [inline test/feature/reverse_routing.feature]: I think your feature specs should all have a basic Express app with Expressive loaded and otherwise empty routes and controllers directories. This should be set up in a global `beforeEach`.

File: test/feature/reverse_routing.feature (hunk 5 above)
Comment (kevgo) [inline test/feature/reverse_routing.feature]: That's a debatable point. I would say that's true if I give it a dedicated name through `as`. If the helper name is derived from the URL, and the helper just provides a safer way to encode URL parameters for example, then it seems plausible that it should change if the URL changes.

File: test/feature/reverse_routing.feature (hunk 6 above)
Comment (kevgo) [inline test/feature/reverse_routing.feature]: I like your version. Sold!

File: test/feature/support/helpers.coffee (hunk 8 above)
Comment (kevgo) [inline test/feature/support/helpers.coffee:23]: :trophy:

File: README.md [hunk 10]
```diff
@@ -43,6 +43,59 @@ Now:
 * visit `/hello` in your browser
 
 
+## Reverse routing
+Exprestive exports reverse routing functions to a `paths` object. By default it is saved to `res.locals.paths` for easy access from views.
```
Comment (kevgo) [inline README.md]: While "reverse routing functions" is technically correct, that's a term that doesn't help users of your library understand what they can use them for. "routing helper functions" or "url builder functions" seems easier understandable. Aha, I can use them to create URLs to point to!

File: test/feature/reverse_routing.feature (hunk 7 above)
Comment (alexdavid) [inline test/feature/reverse_routing.feature]: I was thinking about this, but I think I want to add tests to make sure it throws errors when there is no controllers directory etc..

Would you recommend for those I just delete the controllers directory?

File: README.md (hunk 10 above)
Comment (alexdavid) [inline README.md]: +1  reverse routing is what the Play Framework calls it so I just decided to use that. I like "url building functions"

File: test/feature/reverse_routing.feature (hunk 7 above)
Comment (kevgo) [inline test/feature/reverse_routing.feature]: Yes, for those edge-case specs simply delete the directory. 

In general with Cucumber, this is supposed to be non-technical, almost like human-facing documentation that a PM or an RMS executive understands by just glancing over it. Imagine you explain this to a human idiot (like me). When explaining the different aspects of the routes file to me, you wouldn't always start with an empty directory, and then always first create a controllers directory. You would start with a "normal" app, and then tell me "look, when I add 'GET /some/route' to the routes file here, then it calls this controller automatically." After I say "Wow, cool", and ask what happens if the controller doesn't exist, you would simply delete it, boot the app again, and show me that it throws the right exception. The cucumber specs should be on the same level. Simple, casual, end-user friendly walk through of the different functional aspects. 

They should be like the UI. Wrapping the complex technicalities under an intuitive and easy to use UI. They shouldn't sound/feel/be very technical.

Comment (alexdavid) [inline test/feature/reverse_routing.feature]: Cool, that makes sense. I'll make that change

**Comment by alexdavid** [comment]
Going to merge this since I'm about to block on this and since Exprestive is still in version 0.x we can deal with polishing in a future PR if you think its necessary.
